### ISSUE TITLE
AI-Powered Category Mapping with Learning Loop

### ISSUE THREAD

**Issue by akisma** [akisma/CostFX#36]
**Epic**: Agent Intelligence Enhancement  
**Priority**: P2 - Enhancement  
**Dependencies**: Issue #20 (provides category mapping foundation)

## Product Value

Replace static fuzzy-match category mapping with AI-powered inference that learns from user corrections. This embodies the mission of "making Dave's intuition available programmatically" - the AI learns that "Exotic Fruits" belongs under `produce.fruits.exotic` and "Truffle Oil" goes to `condiments.oils.premium`, continuously improving from every user confirmation.

## Current State (Issue #20)

Static fuzzy matching with ~100 predefined mappings:
```javascript
const mappings = {
  'lemon': 'produce.fruits.citrus',
  'spices': 'spices',
  // ... 100+ predefined mappings
};

// Unmapped categories → null categoryId
// User must manually map each new category
```

**Limitations**:
- Fixed mapping dictionary (requires code updates)
- No learning from user corrections
- Can't handle regional variations or specialty items
- 70-80% coverage, 20-30% needs manual mapping
- No cross-restaurant intelligence

### Three-Tier Resolution Strategy

**Tier 1: Predefined Mappings** (70-80% coverage, instant)
- Fast lookup for common categories
- No API calls, zero latency

**Tier 2: AI Inference** (20-25% coverage, 2-3 seconds)
- AI analyzes category name + existing hierarchy
- Returns suggested path with confidence score
- High confidence (>85%) → auto-apply
- Medium confidence (60-85%) → queue for user review
- Low confidence (<60%) → skip to Tier 3

**Tier 3: User Confirmation** (5% edge cases)
- User reviews and confirms/corrects mapping
- System learns from every confirmation
- Builds training data for future improvements

### AI-Powered Category Inference

```javascript
// AI analyzes category in context
const result = await aiCategoryMapper.inferCategory({
  categoryName: 'Exotic Fruits',
  existingCategories: [
    'produce.fruits',
    'produce.fruits.citrus',
    'produce.fruits.berries'
  ],
  restaurantType: 'fine_dining',
  context: {
    otherCategories: ['Premium Proteins', 'Imported Cheeses']
  }
});

// AI Response:
{
  path: 'produce.fruits.exotic',
  confidence: 0.87,
  reasoning: [
    'Contains word "fruits" - maps to produce.fruits hierarchy',
    'Adjective "exotic" suggests specialty subcategory',
    'Restaurant uses other premium/specialty categories',
    'Path follows existing hierarchy pattern'
  ],
  alternativePaths: [
    { path: 'produce.specialty', confidence: 0.65 },
    { path: 'produce.fruits.imported', confidence: 0.72 }
  ]
}
```

### Learning Loop

```javascript
// Scenario 1: AI gets it right
AI suggests: 'produce.fruits.exotic' (87% confidence)
User confirms: 'produce.fruits.exotic' ✅
System learns: Reinforce this pattern

// Scenario 2: User corrects AI
AI suggests: 'produce.fruits.exotic' (87% confidence)
User corrects: 'produce.specialty.exotic' ✅
System learns: Update model weights, prefer this correction

// Scenario 3: Cross-restaurant learning
Restaurant A: 'Exotic Fruits' → 'produce.fruits.exotic'
Restaurant B: 'Exotic Fruit' → 'produce.fruits.exotic'
Restaurant C sees: 'Exotic Tropical Fruits'
AI learns: High confidence for 'produce.fruits.exotic' (pattern recognition)
```

### Continuous Improvement

```javascript
// System tracks accuracy over time
{
  totalMappings: 1247,
  aiInferences: 312,
  aiCorrect: 289,      // 92.6% accuracy
  userCorrected: 23,   // 7.4% needed correction
  
  learningImpact: {
    week1Accuracy: 0.78,
    week4Accuracy: 0.87,
    week12Accuracy: 0.93,  // Improves over time
    currentAccuracy: 0.926
  }
}
```

### Phase 1: AI Integration (2-3 weeks)

**1.1 AI Provider Setup**
- Support OpenAI (gpt-4o-mini) and Anthropic (claude-3-haiku)
- Configurable via environment variables
- Cost-effective models for high-volume inference

**1.2 Inference Service**
```javascript
class AICategoryMapper {
  async inferCategory(categoryName, context) {
    // Build prompt with existing category hierarchy
    const prompt = this._buildPrompt(categoryName, context);
    
    // Call AI provider
    const response = await this._callAI(prompt);
    
    // Parse and validate response
    const result = this._parseResponse(response);
    
    // Apply confidence thresholds
    if (result.confidence > 0.85) {
      return { action: 'auto_apply', ...result };
    } else if (result.confidence > 0.60) {
      return { action: 'queue_for_review', ...result };
    } else {
      return { action: 'manual_mapping', ...result };
    }
  }
}
```

**1.3 Database Schema**
```sql
-- Store pending confirmations
CREATE TABLE category_mapping_confirmations (
  id SERIAL PRIMARY KEY,
  restaurant_id INTEGER NOT NULL,
  square_category_id VARCHAR(255) NOT NULL,
  category_name VARCHAR(255) NOT NULL,
  ai_suggested_path VARCHAR(255),
  ai_confidence DECIMAL(3,2),
  ai_reasoning TEXT,
  confirmed_path VARCHAR(255),
  status VARCHAR(50) DEFAULT 'pending',
  created_at TIMESTAMPTZ DEFAULT NOW(),
  confirmed_at TIMESTAMPTZ
);

-- Store learning data
CREATE TABLE category_mapping_learning (
  id SERIAL PRIMARY KEY,
  restaurant_id INTEGER NOT NULL,
  category_name VARCHAR(255) NOT NULL,
  mapped_path VARCHAR(255) NOT NULL,
  source VARCHAR(50) NOT NULL, -- 'predefined', 'ai_inference', 'user_confirmed'
  ai_was_correct BOOLEAN, -- Track AI accuracy
  created_at TIMESTAMPTZ DEFAULT NOW()
);
```

### Phase 2: User Interface (1-2 weeks)

**2.1 Import Summary with Warnings**
```javascript
// After import, show unmapped categories
{
  success: true,
  transformed: 347,
  warnings: {
    unmappedCategories: 5,
    needsReview: 3  // AI suggestions pending review
  },
  unmappedCategories: [
    {
      name: 'Exotic Fruits',
      itemCount: 12,
      aiSuggestion: {
        path: 'produce.fruits.exotic',
        confidence: 0.78,
        reasoning: '...'
      }
    }
  ]
}
```

**2.2 Category Confirmation Modal**
- Show Square category name
- Display AI suggestion with confidence %
- Show reasoning for transparency
- Dropdown to confirm or select different path
- Preview affected items
- One-click confirm or manual override

**2.3 Settings: Category Mappings Page**
- Active Mappings tab
- Pending Review tab (with badge count)
- Learning History tab (show AI accuracy trends)
- Bulk approve AI suggestions
- Edit/delete mappings

### Phase 3: Learning & Optimization (2-3 weeks)

**3.1 Model Fine-tuning**
- Collect 1000+ confirmed mappings
- Fine-tune model on restaurant-specific patterns
- A/B test fine-tuned vs base model
- Monitor accuracy improvements

**3.2 Cross-Restaurant Learning**
- Privacy-preserving aggregation
- Learn patterns across all restaurants
- "Restaurants like yours mapped X to Y"
- Federated learning approach

**3.3 Personalization**
- Restaurant-specific model weights
- Learn cuisine-specific patterns (Italian vs Japanese)
- Adapt to regional naming variations
- Remember corrections per restaurant

### Phase 4: Advanced Features (1-2 weeks)

**4.1 Bulk Operations**
- "Approve all AI suggestions over 85% confidence"
- "Review and approve 10 pending mappings"
- Batch confirmation workflow

**4.2 Analytics Dashboard**
- AI accuracy trends over time
- Most common unmapped categories
- Categories needing review
- Cost savings from AI automation

**4.3 Explainability**
- Show AI reasoning for each suggestion
- Confidence breakdown by factor
- Alternative paths considered
- Similar mappings from other restaurants

### Accuracy Metrics
- **AI Inference Accuracy**: >90% correct suggestions
- **Auto-Apply Success Rate**: >95% for high-confidence (>85%)
- **User Correction Rate**: <10% corrections needed
- **Coverage Improvement**: 70% → 95% mapped categories

### Efficiency Metrics
- **Time Savings**: 80% reduction in manual mapping time
- **User Satisfaction**: "Just works" experience
- **Cost Efficiency**: </bin/zsh.01 per category mapping (AI costs)
- **Learning Speed**: Accuracy improves 5% every 100 confirmations

### Business Value Metrics
- **Onboarding Speed**: New restaurants productive in <10 minutes
- **Cross-Restaurant Learning**: Collective intelligence improves all users
- **Scalability**: Support 100+ POS providers without manual mapping
- **User Trust**: High confidence in AI suggestions

### AI Providers
```javascript
// OpenAI (Default)
model: 'gpt-4o-mini'        // Fast, cheap (/bin/zsh.150/1M input tokens)
temperature: 0.3            // Low for consistency
max_tokens: 200             // Short responses

// Anthropic (Alternative)
model: 'claude-3-haiku-20240307'  // Fast, cheap
max_tokens: 200
```

### Prompt Engineering
```
You are a restaurant inventory categorization expert.

Given a category name from a POS system, map it to the most appropriate 
hierarchical category path using the existing category structure.

Existing categories:
{existing_category_tree}

Common restaurant hierarchy:
- produce.fruits.citrus (lemons, limes, oranges)
- produce.vegetables.root (carrots, potatoes)
- proteins.meat.beef (ribeye, sirloin)
- spices.premium (saffron, truffle, vanilla)
... etc

POS Category: "{category_name}"
Restaurant Type: {restaurant_type}
Context: {other_categories}

Return JSON:
{
  "path": "hierarchical.category.path",
  "confidence": 0.0-1.0,
  "reasoning": ["reason 1", "reason 2"]
}

Rules:
- Use existing paths when possible
- Maximum 4 levels deep
- Use snake_case
- Be specific but not overly granular
```

### Cost Analysis
```
Per Category Mapping:
- Input tokens: ~500 (prompt + context)
- Output tokens: ~100 (response)
- Cost per mapping: ~/bin/zsh.0001 (OpenAI gpt-4o-mini)

For 1000 mappings/month:
- Total cost: ~/bin/zsh.10/month
- vs Manual mapping: 1000 * 2 minutes = 33 hours = +
- ROI: 5000x cost savings
```

### Phase 1: Core AI Integration (2-3 weeks)
- [ ] AI provider setup (OpenAI/Anthropic)
- [ ] Inference service implementation
- [ ] Database schema for confirmations/learning
- [ ] Basic API endpoints
- [ ] Unit tests (30+ tests)

### Phase 2: User Interface (1-2 weeks)
- [ ] Import warnings for unmapped categories
- [ ] Category confirmation modal
- [ ] Settings: category mappings page
- [ ] Bulk operations UI
- [ ] Integration tests (15+ tests)

### Phase 3: Learning & Optimization (2-3 weeks)
- [ ] Learning loop implementation
- [ ] Model fine-tuning infrastructure
- [ ] Cross-restaurant aggregation
- [ ] Personalization engine
- [ ] Performance tests

### Phase 4: Advanced Features (1-2 weeks)
- [ ] Analytics dashboard
- [ ] Explainability features
- [ ] Bulk approval workflows
- [ ] Documentation

## Configuration

```bash
# .env

# AI Category Mapping
CATEGORY_MAPPING_AI_ENABLED=true              # Enable AI inference
CATEGORY_MAPPING_AI_PROVIDER=openai           # 'openai' or 'anthropic'
OPENAI_API_KEY=sk-...                         # OpenAI API key
ANTHROPIC_API_KEY=sk-ant-...                  # Anthropic API key
CATEGORY_MAPPING_AI_MIN_CONFIDENCE=0.75       # Auto-apply threshold
CATEGORY_MAPPING_AI_QUEUE_CONFIDENCE=0.60     # Queue for review threshold
CATEGORY_MAPPING_LEARNING_ENABLED=true        # Enable learning loop
```

## Dave's Intuition Made Programmatic

This feature exemplifies our mission: Dave knows instantly that "Exotic Fruits" should be under produce.fruits, and "White Truffle Oil" is a premium condiment. Our AI learns this intuition and makes it available to all restaurants:

- **New restaurant benefit**: Instantly leverages collective wisdom
- **Continuous improvement**: Every confirmation makes the system smarter
- **Cross-restaurant intelligence**: Learn from 1000s of mapping decisions
- **Adapts to each restaurant**: Personalizes to cuisine type and preferences

**Real-World Example**:

The system becomes smarter with every restaurant that joins.

---

**Estimated Effort**: 6-10 weeks (design + implementation + testing + optimization)  
**Value**: Very High - Core differentiator for "AI that understands restaurant operations"  
**ROI**: 5000x cost savings (AI costs vs manual mapping time)

---

## Related Issues
- Issue #20: Square Inventory Synchronization (provides foundation)
- Issue #34: AI-Powered Unit Inference (similar learning pattern)
- Issue #35: AI-Powered Dynamic Variance Thresholds (similar learning pattern)
